pubtools/pulplib/_impl/model/repository/file.py

```diff
@@ -20,3 +23,53 @@ class FileRepository(Repository):
     )
 
     mutable_urls = attr.ib(default=attr.Factory(lambda: ["PULP_MANIFEST"]), type=list)
+
+    def upload_file(self, filename, relative_url=None):
+        """Upload a file to this repository.
+
+        The upload operation includes 4 steps:
```

Comment:
Actually, can you please *not* document it to this level of detail?  Sorry, I should have caught that on earlier review.

It's better if the doc doesn't specify exactly what steps are needed.  If you document that, then it can be argued we can't change it without breaking backwards-compatibility.  After all, you documented that it creates and deletes an upload, so it would be reasonable for someone to somehow rely on that.  That would block us from implementing behavior in this API such as: if file already exists in same repo with same name and checksum, skip the upload.

I think the docs should mostly focus on the postcondition of the method.  I'd put it like: Ensure this repository has a file at relative_url containing the content of filename.